src/stream/src/executor/top_n_new.rs

```diff
@@ -0,0 +1,904 @@
+                            // has no limit, emit to result set
+                            new_ops.push(Op::Insert);
+                            new_rows.push(row.clone());
+                            continue;
+                        }
+
+                        // check whether new key in the [0, limit]
+                        if let Some(end_key) = end_row.ordered_key.as_ref() {
+                            if ordered_pk_row <= *end_key {
+                                // update boundary if needed
+                                let (_, new_end_row) = self
+                                    .managed_state
+                                    .find_range(self.offset, self.limit, epoch)
+                                    .await?;
+
+                                if end_row.ordered_key != new_end_row.ordered_key {
+                                    new_ops.push(Op::Delete);
+                                    new_rows.push(end_row.row);
+                                }
+
+                                new_ops.push(Op::Insert);
+                                new_rows.push(row.clone());
+                            }
+                        }
+                    } else if self.offset + 1 == self.managed_state.total_count() {
+                        // handle the boundary case
+                        let (new_start_row, _) = self
+                            .managed_state
+                            .find_range(self.offset, self.limit, epoch)
+                            .await?;
+                        new_ops.push(Op::Insert);
+                        new_rows.push(new_start_row.row);
+                    } else if let Some(start_key) = start_row.ordered_key.as_ref() {
+                        // Insert into [0, offset), the topn range will move backward
+                        if ordered_pk_row < *start_key {
+                            let (new_start_row, new_end_row) = self
+                                .managed_state
+                                .find_range(self.offset, self.limit, epoch)
+                                .await?;
+
+                            // update boundary
+                            if new_start_row.is_valid() {
+                                new_ops.push(Op::Insert);
+                                new_rows.push(new_start_row.row.clone());
+                            }
+                            if end_row.ordered_key != new_end_row.ordered_key && end_row.is_valid()
+                            {
+                                assert!(new_end_row.ordered_key <= end_row.ordered_key);
+                                new_ops.push(Op::Delete);
+                                new_rows.push(end_row.row);
+                            }
+                        } else if !end_row.is_valid() {
+                            // no limit
+                            new_ops.push(Op::Insert);
+                            new_rows.push(row);
```

Comment:
re-use a function on this?

Reply:
Do you mean wrap those two lines to a function?  That will make the code cleaner, but I think current way is also readable. 
```
new_ops.push(Op::Insert);
new_rows.push(row);
```